### ISSUE TITLE
Export history trail as GPX

### ISSUE THREAD

**Issue by moving-bits** [cgeo/cgeo#8424]
Taken from a discussion in #8416:

> BTW: Would it be a big deal to implement an export of the history line track as gpx?
> Would be a perfect helper for OSM edits ...

Shouldn't be too hard to implement. But needs to be accompanied by a larger trail history length - currently history trail gets cut off after 700 entries, which is memory saving, but by far not enough for a day trip. (That needs to be made configurable probably, so that someone with enough RAM can set this to a higher value.)

**Comment by MagpieFourtyTwo** [comment]
Memory? Ok ... I had assumed that the history is saved to the database, as it survives app restarts ...

Anyway - for at least creating a template for OSM edits, the current length is more than sufficient. Even two or three kilometers would be enough, as most of the trails are already mapped anyway and all you need (for at least this purpose) are just the non- or faulty mapped parts ... and in the rare cases where they are still longer, one can simply export multiple times, I would say ...  ;)

Nevertheless, a longer history trail history would of course be nice, but regarding memory needs (I guess, at least timestamp (4 byte) and 2 reals (2*4 bytes), thus 12 bytes minimum per point, in sum 1,2 MB for 10.000 points, which is indeed a lot, considering Android's max of 16 MB) this may require mayor reworks ...?

So, as for me, just the "simple" export would be good enough, as a first approach.  ;)

**Comment by MagpieFourtyTwo** [comment]
Just skimmed the source code changes of today's commits ...
Went a little deeper here and there ...
WOW.
If I would have known the scope of necessary changes, I wouldn't have asked for them, but would just continue to draw tracks from my memory resp. still just toggle `show/hide history line` to get aware of the trail type underneath ...  ;)
And again: THX!  :))

**Comment by Lineflyer** [comment]
> And again: THX! :))

BTW: As a thank you, you could help us by testing this feature and reporting back any problems you see. I personally do not use the trail and it would cost me much effort to test this. This would be very kind.

**Comment by MagpieFourtyTwo** [comment]
Tried to import a c:geo-exported track into [osm.org](https://www.openstreetmap.org/traces), Garmin's [Basecamp](https://www.garmin.com/en-US/software/basecamp) and the Android app of [kurviger.de](https://kurviger.de) (BTW: the first and latter can be used for testing by everyone with no hassle).

Result was - sorry - complete fail:
- osm.org did not even show an error message, but the tracks do not show up
- kurviger.de just refused to load
- Basecamp just said "Error opening file"

While digging into the text file Notepad++ (with `XML Tools` plugin installed) said:

`"XML Parsing error at line 1: Attribute n965:lat redefined"`

After pretty printing the one lined XML stuff it's in fact s line 10, i. e. the first `trkpt` tag:

`<trkpt n0:lat="48.4711" n0:lat="9.4711" xmlns:n0="http://www.topografix.com/GPX/1/0" />`

Regretfully even after changing the second `lat` to `lon` Notepad++ was still the only one which gave further error messages (can manually be triggered via menu `Plugins` -> `XML Tools` -> `Validate now`):

`ERROR: Element '{http://www.topografix.com/GPX/1/0}metadata': This element is not expected. Expected is one of ( {http://www.topografix.com/GPX/1/0}name, {http://www.topografix.com/GPX/1/0}desc, {http://www.topografix.com/GPX/1/0}author, {http://www.topografix.com/GPX/1/0}email, {http://www.topografix.com/GPX/1/0}url, {http://www.topografix.com/GPX/1/0}urlname, {http://www.topografix.com/GPX/1/0}time, {http://www.topografix.com/GPX/1/0}keywords, {http://www.topografix.com/GPX/1/0}bounds, {http://www.topografix.com/GPX/1/0}wpt ).
`

Don't know if this tag is really needed (esp. as Basecamp exports do not even contain anything like this), but the error could be "solved" by just deleting the entire `xmlns:n0="http://www.topografix.com/GPX/1/0"` part completely.

This slightly stripped file could ...
- finally be imported into Basecamp, track is displayed perfectly fine
- Kurviger.de: looks good, now. imports and displays the stripped file.
- OSM: after uploading the track is listed as `Pending` (takes up to half an hour until they have put it in their db), but after a while the entry is completely gone, no results or error messages at all ... nothing. Strange. Even exported an OSM-track and skimmed it, but did not see any major differences - perhaps you may have a look into one of these?  Their tracks can be downloaded [here](https://www.openstreetmap.org/traces).

**Comment by MagpieFourtyTwo** [comment]
Just got feedback from OSM - vial Mail. About 36 hours after import ... <w/o words>

Ok, the initial c:geo GPX (as is from c:geo, i. e. containing even the erroneous redefinition of `lat`) has been rejected because of:
```
Hi,

It looks like your GPX file trail_2020_06_17_06_43_21.gpx with the description test and no tags. failed to import. Here is the error:

Fatal error: Attribute n0:lat redefined at :1. /srv/www.openstreetmap.org/rails/config/initializers/libxml.rb:4:in 'block in <top (required)>' /srv/www.openstreetmap.org/rails/lib/gpx.rb:18:in 'read' /srv/www.openstreetmap.org/rails/lib/gpx.rb:18:in 'parse_file' /srv/www.openstreetmap.org/rails/lib/gpx.rb:59:in 'rescue in points' /srv/www.openstreetmap.org/rails/lib/gpx.rb:47:in 'points' /srv/www.openstreetmap.org/rails/app/models/trace.rb:254:in 'each' /srv/www.openstreetmap.org/rails/app/models/trace.rb:254:in 'each_slice' /srv/www.openstreetmap.org/rails/app/models/trace.rb:254:in 'import' /srv/www.openstreetmap.org/rails/app/jobs/trace_importer_job.rb:5:in 'perform' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/execution.rb:40:in 'block in perform_now' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:112:in 'block in run_callbacks' /var/lib/gems/2.5.0/gems/i18n-0.9.5/lib/i18n.rb:268:in 'with_locale' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/translation.rb:9:in 'block (2 levels) in <module:Translation>' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in 'instance_exec' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in 'block in run_callbacks' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/core_ext/time/zones.rb:66:in 'use_zone' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/timezones.rb:9:in 'block (2 levels) in <module:Timezones>' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in 'instance_exec' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in 'block in run_callbacks' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/logging.rb:25:in 'block (4 levels) in <module:Logging>' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/notifications.rb:180:in 'block in instrument' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/notifications/instrumenter.rb:24:in 'instrument' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/notifications.rb:180:in 'instrument' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/logging.rb:24:in 'block (3 levels) in <module:Logging>' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/logging.rb:45:in 'block in tag_logger' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/tagged_logging.rb:80:in 'block in tagged' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/tagged_logging.rb:28:in 'tagged' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/tagged_logging.rb:80:in 'tagged' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/logging.rb:45:in 'tag_logger' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/logging.rb:21:in 'block (2 levels) in <module:Logging>' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in 'instance_exec' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in 'block in run_callbacks' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:139:in 'run_callbacks' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/execution.rb:39:in 'perform_now' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/execution.rb:25:in 'block in execute' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:112:in 'block in run_callbacks' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/railtie.rb:43:in 'block (4 levels) in <class:Railtie>' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/execution_wrapper.rb:88:in 'wrap' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/reloader.rb:72:in 'block in wrap' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/execution_wrapper.rb:88:in 'wrap' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/reloader.rb:71:in 'wrap' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/railtie.rb:42:in 'block (3 levels) in <class:Railtie>' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in 'instance_exec' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in 'block in run_callbacks' /var/lib/gems/2.5.0/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:139:in 'run_callbacks' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/execution.rb:23:in 'execute' /var/lib/gems/2.5.0/gems/activejob-6.0.3.2/lib/active_job/queue_adapters/delayed_job_adapter.rb:42:in 'perform' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/backend/base.rb:81:in 'block in invoke_job' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:61:in 'block in initialize' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:66:in 'execute' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:40:in 'run_callbacks' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/backend/base.rb:78:in 'invoke_job' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:230:in 'block (2 levels) in run' /usr/lib/ruby/2.5.0/timeout.rb:93:in 'block in timeout' /usr/lib/ruby/2.5.0/timeout.rb:103:in 'timeout' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:230:in 'block in run' /usr/lib/ruby/2.5.0/benchmark.rb:308:in 'realtime' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:229:in 'run' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:312:in 'block in reserve_and_run_one_job' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:61:in 'block in initialize' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:66:in 'execute' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:40:in 'run_callbacks' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:312:in 'reserve_and_run_one_job' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:213:in 'block in work_off' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:212:in 'times' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:212:in 'work_off' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:175:in 'block (4 levels) in start' /usr/lib/ruby/2.5.0/benchmark.rb:308:in 'realtime' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:174:in 'block (3 levels) in start' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:61:in 'block in initialize' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:66:in 'execute' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:40:in 'run_callbacks' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:173:in 'block (2 levels) in start' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:172:in 'loop' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:172:in 'block in start' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/plugins/clear_locks.rb:7:in 'block (2 levels) in <class:ClearLocks>' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:79:in 'block (2 levels) in add' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:61:in 'block in initialize' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:79:in 'block in add' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:66:in 'execute' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/lifecycle.rb:40:in 'run_callbacks' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/worker.rb:171:in 'start' /var/lib/gems/2.5.0/gems/delayed_job-4.1.8/lib/delayed/tasks.rb:9:in 'block (2 levels) in <top (required)>' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/task.rb:281:in 'block in execute' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/task.rb:281:in 'each' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/task.rb:281:in 'execute' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/task.rb:219:in 'block in invoke_with_call_chain' /usr/lib/ruby/2.5.0/monitor.rb:226:in 'mon_synchronize' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/task.rb:199:in 'invoke_with_call_chain' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/task.rb:188:in 'invoke' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/application.rb:160:in 'invoke_task' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/application.rb:116:in 'block (2 levels) in top_level' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/application.rb:116:in 'each' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/application.rb:116:in 'block in top_level' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/application.rb:125:in 'run_with_threads' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/application.rb:110:in 'top_level' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/application.rb:83:in 'block in run' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/application.rb:186:in 'standard_exception_handling' /var/lib/gems/2.5.0/gems/rake-13.0.1/lib/rake/application.rb:80:in 'run' /var/lib/gems/2.5.0/gems/rake-13.0.1/exe/rake:27:in '<top (required)>' /usr/local/bin/rake:23:in 'load' /usr/local/bin/rake:23:in '<main>'

More information about GPX import failures and how to avoid them can be found at: https://wiki.openstreetmap.org/wiki/GPX_Import_Failures

OpenStreetMap
```

And regarding the edited one (with corrected `lon` and without `xmlns` tag), which has been imported and displayed by Basecamp and kurviger.de:
```
Hi,

It looks like your GPX file trail_2020_06_17_06_43_21_basecamp_ok.gpx with the description test - basecamp ok and no tags. failed to import. Here is the error:

0 points parsed ok. Do they all have lat,lng,alt,timestamp?

More information about GPX import failures and how to avoid them can be found at: https://wiki.openstreetmap.org/wiki/GPX_Import_Failures

OpenStreetMap
```

Seems as if OSM additionally requires altitude and timestamp.

**Comment by moving-bits** [comment]
@MagpieFourtyTwo  Thanks for testing!

> `"XML Parsing error at line 1: Attribute n965:lat redefined"`

Oops. C&P error... Corrected with #8470.

> After pretty printing the one lined XML stuff it's in fact s line 10, i. e. the first `trkpt` tag:
> `<trkpt n0:lat="48.4711" n0:lat="9.4711" xmlns:n0="http://www.topografix.com/GPX/1/0" />`

Don't know why the XMLSerializer inserts individual namespace information here, although everything around is already in this namespace. - I found a way to actively suppress this behavior here, fixed in #8470.

> `ERROR: Element '{http://www.topografix.com/GPX/1/0}metadata':

Looks like it's supported with `GPX/1/1` only - changed it with #8470 as well.

> Seems as if OSM additionally requires altitude and timestamp.

This is a different issue, not regarding the export itself (it should be a valid GPX track file now), but using the exported file for a specific service which requires additional data (which we don't have yet). Let's open a separate issue for that.
